---
title: "FLOE Wp0 vs Wp0.3"
author: "Christina Chen"
date: "2024-04-29"
output: html_document
---

## Set Up

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load libraries}
library(tidyverse)
library(devtools)
library(edgeR)
```

```{r read files}
sdiv <- read_tsv("../input/Christina_HiFi_Sdiv_Gene_Counts.tsv")
sample <- read_csv("../input/Hydrothermal_Round_1_Sample_Descriptions.csv")
```

# Wp0 vs Wp0.3 Comparison

## Data

```{r retain samples we want}
compare_sample <- sample %>%
  filter(condition == "24 hr 20˚ L wp0" | condition == "24 hr 20˚ L wp-0.3") %>%
  filter(population == "CAAN1" | population == "STDI" | population == "STGL2" | population == "STGL3" | population == "STIN" | population == "STPO1" | population == "STTO-BH") %>%
  filter(sample != "a12") %>%
  filter(sample != "i9b") %>%
  filter(sample != "k15")
compare_sample
```

```{r retain counts we want}
compare_sdiv <- sdiv %>%
  select(1, compare_sample$sample) %>%
  rename("gene_id" = 1) %>%
  select(sort(colnames(.))) %>%
  select(gene_id, everything())
compare_sdiv
```

```{r checking names match}
names(compare_sdiv[,-1]) == compare_sample$sample
```

## Cor Table

```{r doing cor()}
compare_sdiv %>%
  select(-gene_id) %>%
  cor() %>%
  gplots::heatmap.2(Rowv=FALSE, Colv=FALSE, dendrogram="none", trace = "none", col=viridis::viridis(25, begin=.25), margins=c(7,8))
# the f and m populations seem to show clear differences between light and dark
# the i and k populations are less distinctive
```

## Calculate Normalization Factors

```{r making a matrix}
compare_sdiv_matrix <- compare_sdiv %>%
  select(-gene_id) %>%
  as.matrix()
rownames(compare_sdiv_matrix) <- compare_sdiv$gene_id
```

```{r calculating normalization factors}
dge_data <- DGEList(counts = compare_sdiv_matrix,
                    group = compare_sample$group,
                    samples = compare_sample)

dge_data <- calcNormFactors(dge_data, method = "TMM")
dge_data$samples
```

## Plotting BCV

```{r plotting BCV}
plotMDS(dge_data)
```

## Extract Normalized Data

```{r extract normalized data}
compare_sdiv_normal <- cpm(dge_data)
compare_sdiv_normal_log <- cpm(dge_data, log = T)
```

```{r boxplots of normalized data}
boxplot(compare_sdiv_normal)
boxplot(compare_sdiv_normal_log)
# to be removed: i07, i08, i10, i11, j06, j09, j10, k07, k09, k12, l06, l09, l13, m07, m12, m13, m14, m15
```

## Removal of Non-normalized

```{r removing the non-normalized}
to_remove <- c("i07", "i08", "i10", "i11", "j06", "j09", "j10", "k07", "k09", "k12", "l06", "l09", "l13", "m07", "m12", "m13", "m14", "m15")
compare_sample <- compare_sample %>%
  filter(!(sample %in% to_remove))
compare_sample
```

```{r removing the not-normalized too}
compare_sdiv <- compare_sdiv %>%
  select(-to_remove)
compare_sdiv
```

```{r checking names match again}
names(compare_sdiv[,-1]) == compare_sample$sample
```

## Cor Table Again

```{r doing cor() again}
compare_sdiv %>%
  select(-gene_id) %>%
  cor() %>%
  gplots::heatmap.2(Rowv=FALSE, Colv=FALSE, dendrogram="none", trace = "none", col=viridis::viridis(25, begin=.25), margins=c(7,8))
# looks better, but i and j and k populations are still mixed
```

## Calculate Normalization Factors Again

```{r making a matrix again}
compare_sdiv_matrix <- compare_sdiv %>%
  select(-gene_id) %>%
  as.matrix()
rownames(compare_sdiv_matrix) <- compare_sdiv$gene_id
```

```{r calculating normalization factors again}
dge_data <- DGEList(counts = compare_sdiv_matrix,
                    group = compare_sample$group,
                    samples = compare_sample)

dge_data <- calcNormFactors(dge_data, method = "TMM")
dge_data$samples
```

## Plotting BCV again

```{r plotting BCV again}
plotMDS(dge_data)
```

## Extract Normalized Data Again

```{r extract normalized data again}
compare_sdiv_normal <- cpm(dge_data)
compare_sdiv_normal_log <- cpm(dge_data, log = T)
```

```{r boxplots of normalized data again}
boxplot(compare_sdiv_normal)
boxplot(compare_sdiv_normal_log)
```

## Calculate Dispersion Factors & Estimate Dispersions

```{r make a model matrix}
design <- model.matrix(~ population, data = compare_sample)
rownames(design) <- compare_sample$sample.description
design
```

```{r estimate dispersion: common}
dge_data <- estimateGLMCommonDisp(dge_data, design, verbose = TRUE)
# Disp = 0.17332 , BCV = 0.4163
```

```{r estimate dispersion: trended}
dge_data <- estimateGLMTrendedDisp(dge_data, design)
```

```{r estimate dispersion: tagwise}
dge_data <- estimateGLMTagwiseDisp(dge_data, design)
```

```{r plotting biological coefficent of variation and average log CPM}
plotBCV(dge_data)
```

## Find Differentially Expressed Genes

```{r full model}
fit <- glmFit(dge_data, design)
```

```{r finding differentially expressed genes again}
compare_lrt <- glmLRT(fit, coef = colnames(design[,-1]))
```

```{r view the differentially expressed genes again}
topTags(compare_lrt)
```

```{r summarise number of differentially expressed genes again}
summary(decideTestsDGE(compare_lrt, p.value=0.01))
```

```{r extract genes with a FDR < 0.01 (could also use 0.05) and save to a file}
DEgene_LD_compare <- topTags(compare_lrt, n = Inf, p.value = 0.01)$table
write.csv(DEgene_LD_compare,"../output/DEgenes_WP_compare.csv")
```

```{r or if you want to keep all of them}
DEgene_LD_compare_all <- topTags(compare_lrt, n = Inf, p.value = 1)$table
write.csv(DEgene_LD_compare_all,"../output/DEgenes_WP_compare_all.csv")
```

## Box Plot Function

```{r box plot function}
plotDE <- function(genes, dge, sample.description) {
  require(ggplot2)
  tmp.data <- t(log2(cpm(dge[genes,])+1))
  tmp.data <- tmp.data %>%
    as.data.frame() %>%
    rownames_to_column("sample") %>%
    left_join(sample.description,by="sample")
  tmp.data <- tmp.data %>%
    pivot_longer(cols=starts_with("Sdiv"), values_to = "log2_cpm", names_to = "gene")
  pl <- ggplot(tmp.data,aes(x=population,y=log2_cpm,fill=condition))
  pl <- pl + facet_wrap( ~ gene)
  pl <- pl + ylab("log2(cpm)") + xlab("genotype")
  pl <- pl + geom_boxplot()
  pl + theme(axis.text.x  = element_text(angle=45, vjust=1,hjust=1))
}
```

## Looking at FLOEs

```{r are they significant}
DEgene_LD_compare_all %>%
  rownames_to_column("rowname") %>%
  filter(rowname == "Sdiv_ptg000004l_1566-R" | 
         rowname == "Sdiv_ptg000005l_1254-R" | 
         rowname == "Sdiv_ptg000009l_0928-R" | 
         rowname == "Sdiv_ptg000010l_1994-R" | 
         rowname == "Sdiv_ptg000001l_2274-R" | 
         rowname == "Sdiv_ptg000013l_FLOE3-R")
```

```{r using box plot function}
# FLOE1
plotDE(c("Sdiv_ptg000004l_1566-R", "Sdiv_ptg000005l_1254-R"), dge_data, compare_sample)

# FLOE2
plotDE(c("Sdiv_ptg000009l_0928-R", "Sdiv_ptg000010l_1994-R"), dge_data, compare_sample)

# FLOE3
plotDE(c("Sdiv_ptg000001l_2274-R", "Sdiv_ptg000013l_FLOE3-R"), dge_data, compare_sample)
```