---
title: "FLOE notes v3"
author: "Christina Chen"
date: "2024-03-24"
output: html_document
---

# Set Up

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load libraries}
library(tidyverse)
library(devtools)
library(edgeR)
library(shiny)
```

```{r read files}
sdiv <- read_tsv("../input/Christina_HiFi_Sdiv_Gene_Counts.tsv")
sample <- read_csv("../input/Hydrothermal_Round_1_Sample_Descriptions.csv")
```

## Environment Information

- sdiv = counts tsv file that was given, and later with non-normalized removed
- sample = sample cvs file that was given
- dry_sample = sample csv with only dry condition, and without i02 and i03 and non-normalized
- dry_sdiv = counts tsv with only dry condition
- dry_sdiv_matrix = made into a matrix for rownames
- dge_data = DGEList and calcNormFactors
- normal_dry_sdiv = cpm of dge_data
- normal_dry_sdiv_log = cpm of dge_data
- sample_names = names of samples with names(dry_sdiv)
- design = matrix of design conditions
- fit = glmFit of dge_data and design
- dry_lrt = glmLRT of fit and coefficient of all populations
- DEgene_dry = genes with a FDR < 0.01
- DEgene_dry_all = all genes

## Data

```{r retaining dry sdiv ones that have been normalized}
dry_sdiv <- sdiv %>%
  select(1, contains( "01") | contains("02") | contains("03") | contains("04") | contains("05")) %>%
  rename("gene_id" = 1) %>%
  select(-a01, -b02, -b04, -c01, -c02, -c03, -c05, -d01, -d02, -d03, -d04, -e03, -g01, -g02, -g03, -g04, -g05, -j03, -k01, -k03, -k04, -k05, -l01, -l04) %>%
  select(sort(colnames(.))) %>%
  select(gene_id, everything())
dry_sdiv
# ended up figuring out that i could do column_to_rownames()
```

```{r retaining sample descriptions to match dry_sdiv}
dry_sample <- sample %>%
  filter(condition == "Dry") %>%
  filter(sample != "i02",
         sample != "i03",
         sample != "a01",
         sample != "b02",
         sample != "b04",
         sample != "c01",
         sample != "c02",
         sample != "c03",
         sample != "c05",
         sample != "d01",
         sample != "d02",
         sample != "d03",
         sample != "d04",
         sample != "e03",
         sample != "g01",
         sample != "g02",
         sample != "g03",
         sample != "g04",
         sample != "g05",
         sample != "j03",
         sample != "k01",
         sample != "k03",
         sample != "k04",
         sample != "k05",
         sample != "l01",
         sample != "l04")
dry_sample
# how to filter with a c()?
```

```{r different way to do it}
to_remove = c("i02", "i03", "a01", "b02", "b04", "c01", "c02", "c03", "c05", "d01", "d02", "d03", "d04", "e03", "g01", "g02", "g03", "g04", "g05", "j03", "k01", "k03", "k04", "k05", "l01", "l04")
sample %>%
  filter(condition == "Dry") %>%
  filter(!(sample %in% to_remove))
```

```{r checking names match}
names(dry_sdiv[,-1]) == dry_sample$sample
```

## Cor Table

```{r doing cor()}
dry_sdiv %>%
  select(-gene_id) %>%
  cor() %>%
  gplots::heatmap.2(Rowv=FALSE, Colv=FALSE, dendrogram="none", trace = "none", col=viridis::viridis(25, begin=.25), margins=c(7,8))
```

## Calculate Normalization Factors

```{r making a matrix}
dry_sdiv_matrix <- dry_sdiv %>%
  select(-gene_id) %>%
  as.matrix()
rownames(dry_sdiv_matrix) <- dry_sdiv$gene_id
```

```{r calculating normalization factors}
dge_data <- DGEList(counts = dry_sdiv_matrix,
                    group = dry_sample$group,
                    samples = dry_sample)

dge_data <- calcNormFactors(dge_data, method = "TMM")
dge_data$samples
```

## Plotting BCV

```{r plotting BCV}
plotMDS(dge_data)
# why does it look so different from v2?
```

## Extract Normalized Data

```{r extract normalized data}
dry_sdiv_normal <- cpm(dge_data)
dry_sdiv_normal_log <- cpm(dge_data, log = T)
```

```{r boxplots of normalized data}
boxplot(dry_sdiv_normal)
boxplot(dry_sdiv_normal_log)
```

## Calculate Dispersion Factors & Estimate Dispersions

```{r make a model matrix}
design <- model.matrix(~ population, data = dry_sample)
rownames(design) <- dry_sample$sample.description
design
```

```{r estimate dispersion: common}
dge_data <- estimateGLMCommonDisp(dge_data, design, verbose = TRUE)
# Disp = 0.24255 , BCV = 0.4925 
```

```{r estimate dispersion: trended}
dge_data <- estimateGLMTrendedDisp(dge_data, design)
```

```{r estimate dispersion: tagwise}
dge_data <- estimateGLMTagwiseDisp(dge_data, design)
```

```{r plotting biological coefficent of variation and average log CPM}
plotBCV(dge_data)
# trend looks better
```

## Find Differentially Expressed Genes

```{r full model}
fit <- glmFit(dge_data, design)
```

```{r finding differentially expressed genes again}
dry_lrt <- glmLRT(fit, coef = colnames(design[,-1]))
```

```{r view the differentially expressed genes again}
topTags(dry_lrt)
```

### What topTags Shows

- logFC = log2 fold-change in expression, so populationCAAN2 (gene 20946) has 1.31, which means that gene is expressed 2^1.31 (or 2.48) more. if it is  a negative, it is less.
      - between intercept and a model without each population?
- logCPM = average expression across all samples
- LR = likeihood ratio (L(full model)/L(small model))
- PValue = unadjusted p value
- FDR = false discovery rate (p value adjusted for multiple testing)

```{r summarise number of differentially expressed genes again}
summary(decideTestsDGE(dry_lrt, p.value=0.01))
```

```{r extract genes with a FDR < 0.01 (could also use 0.05) and save to a file}
DEgene_dry <- topTags(dry_lrt, n = Inf, p.value = 0.01)$table # what is n = Inf for
write.csv(DEgene_dry,"../output/DEgenes_dry.csv")
```

```{r or if you want to keep all of them}
DEgene_dry_all <- topTags(dry_lrt, n = Inf, p.value = 1)$table
write.csv(DEgene_dry_all,"../output/DEgenes_dry_all.csv")
```

## Box Plot Function

```{r box plot function}
plotDE <- function(genes, dge, sample.description) {
  require(ggplot2)
  tmp.data <- t(log2(cpm(dge[genes,])+1))
  tmp.data <- tmp.data %>%
    as.data.frame() %>%
    rownames_to_column("sample") %>%
    left_join(sample.description,by="sample")
  tmp.data <- tmp.data %>%
    pivot_longer(cols=starts_with("Sdiv"), values_to = "log2_cpm", names_to = "gene")
  pl <- ggplot(tmp.data,aes(x=population,y=log2_cpm))
  pl <- pl + facet_wrap( ~ gene)
  pl <- pl + ylab("log2(cpm)") + xlab("genotype")
  pl <- pl + geom_boxplot()
  pl + theme(axis.text.x  = element_text(angle=45, vjust=1,hjust=1))
}
```

## Looking at FLOEs

```{r are they significant}
DEgene_dry_all %>%
  rownames_to_column("rowname") %>%
  filter(rowname == "Sdiv_ptg000004l_1566-R" | rowname == "Sdiv_ptg000005l_1254-R" | rowname == "Sdiv_ptg000009l_0928-R" | rowname == "Sdiv_ptg000010l_1994-R" | rowname == "Sdiv_ptg000001l_2274-R" | rowname == "Sdiv_ptg000013l_FLOE3-R")
```


```{r using box plot function}
# FLOE1
plotDE(c("Sdiv_ptg000004l_1566-R", "Sdiv_ptg000005l_1254-R"), dge_data, dry_sample)
# FLOE2
plotDE(c("Sdiv_ptg000009l_0928-R", "Sdiv_ptg000010l_1994-R"), dge_data, dry_sample)
# FLOE3
plotDE(c("Sdiv_ptg000001l_2274-R", "Sdiv_ptg000013l_FLOE3-R"), dge_data, dry_sample)
```

## Correlation

```{r reading germination data}
# this is the code given to me
germData  <- read_csv("../input/Dry_and_Wp0_Germination_Data.csv") %>%
  group_by(population) %>%
  mutate(cumulative_prop_germ = max(cumulative_prop_germ)) %>%
  ungroup() %>%
  filter(condition == "Dry") %>%
  select(sample, population, cumulative_prop_germ)
germData
```

```{r combine files}
# I had some issues with figuring out what values I should be using for expression levels
# here I made a new object called cor_floe, where I took the the logFC counts for the FLOE genes and joined the germData
# I wasn't sure if this was correct, since all of these are in relation to CAAN1 population, also there are duplicates since the germData had 5 replicates for each population.
# I also moved on to graphing without averaging out the samples by population
# I tried to graph it in the next code chunk
cor_floe <- DEgene_dry_all %>%
  rownames_to_column("rowname") %>%
  filter(rowname == "Sdiv_ptg000004l_1566-R" | 
         rowname == "Sdiv_ptg000005l_1254-R" | 
         rowname == "Sdiv_ptg000009l_0928-R" | 
         rowname == "Sdiv_ptg000010l_1994-R" | 
         rowname == "Sdiv_ptg000001l_2274-R" | 
         rowname == "Sdiv_ptg000013l_FLOE3-R")

colnames(cor_floe) <- str_remove(colnames(cor_floe), "logFC.population")

cor_floe <- cor_floe %>%
  rename(STTO_BH = STTO.BH) %>%
  pivot_longer(cols = CAAN2:STTO_BH,
               names_to = "population", 
               values_to = "logFC") %>%
  full_join(germData,
            by = "population")

cor_floe
```

```{r correlation between floe expression and germination}
# this is where I graphed it, but I'm not sure how you derived information from this graph, since it's in comparison to CAAN1
# I put lines in to show which colors/genes are which
cor_floe %>%
  ggplot(aes(x = cumulative_prop_germ, y = logFC, color = rowname)) +
  geom_point() +
  geom_line()
```

```{r}
# I also tried to do a correlation heat map, but I don't think this is was what we're going for
cor_floe2 <- DEgene_dry_all %>%
  rownames_to_column("rowname") %>%
  filter(rowname == "Sdiv_ptg000004l_1566-R" | 
         rowname == "Sdiv_ptg000005l_1254-R" | 
         rowname == "Sdiv_ptg000009l_0928-R" | 
         rowname == "Sdiv_ptg000010l_1994-R" | 
         rowname == "Sdiv_ptg000001l_2274-R" | 
         rowname == "Sdiv_ptg000013l_FLOE3-R")

colnames(cor_floe2) <- str_remove(colnames(cor_floe2), "logFC.population")

cor_floe2 %>%
  select(-rowname, -logCPM, -LR, -PValue, -FDR) %>%
  cor() %>%
  gplots::heatmap.2(Rowv=FALSE, Colv=FALSE, dendrogram="none", trace = "none", col=viridis::viridis(25, begin=.25), margins=c(7,8))
```

```{r}
# next, I used cpm() and log2( + 1) on the counts data, kept only the FLOE genes, and averaged out the replicates.
# I then averaged out the germData and joined them together
dry_sdiv_normal_log2 <- log2(cpm(dge_data) + 1)

cor_floe2 <- dry_sdiv_normal_log2 %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  filter(rowname == "Sdiv_ptg000004l_1566-R" | 
         rowname == "Sdiv_ptg000005l_1254-R" | 
         rowname == "Sdiv_ptg000009l_0928-R" | 
         rowname == "Sdiv_ptg000010l_1994-R" | 
         rowname == "Sdiv_ptg000001l_2274-R" | 
         rowname == "Sdiv_ptg000013l_FLOE3-R") %>%
  group_by(rowname) %>%
  mutate(CAAN1 = mean(c(a02, a03, a04, a05)),
         CAAN2 = mean(c(b01, b03, b05)),
         CACO1 = mean(c(c04)),
         CAIN2 = mean(c(d05)),
         STBR3 = mean(c(e01, e02, e04, e05)),
         STDI = mean(c(f01, f02, f03, f04, f05)),
         STGL1 = mean(c(h01, h02)),
         STGL2 = mean(c(i01, i04, i05)),
         STGL3 = mean(c(j01, j02, j04, j05)),
         STIN = mean(c(k02)),
         STPO1 = mean(c(l02, l03, l05)),
         STTO.BH = mean(c(m01 ,m02, m03, m04, m05))) %>%
  select(rowname, CAAN1, CAAN2, CACO1, CAIN2, STBR3, STDI, STGL1, STGL2, STGL3, STIN, STPO1, STTO.BH) %>%
  pivot_longer(cols = CAAN1:STTO.BH,
               names_to = "population",
               values_to = "log2cpm")

germData2 <- germData %>%
  group_by(population) %>%
  summarise(cumulative_prop_germ = mean(cumulative_prop_germ))

cor_floe2 <- cor_floe2 %>%
  full_join(germData2)
  
cor_floe2
```

```{r}
# then i graphed the proportion of germination against the log2cpm
# ptg13l looks bad since we don't have the whole picture with that one
# the lines are there to show them apart
# I'm not sure if this was what we're trying to look for, especially since I'm not seeing any correlation
# You can kind of see that lower log2cpm has higher germination in graph 2, but I'm not sure if that is significant
cor_floe2 %>%
  ggplot(aes(x = cumulative_prop_germ, y = log2cpm, color = rowname)) +
  geom_point() + 
  geom_line()

cor_floe2 %>%
  ggplot(aes(x = cumulative_prop_germ, y = log2cpm)) +
  geom_point() + 
  geom_smooth(method = lm, alpha = 0)
```




