---
title: "FLOE notes v3"
author: "Christina Chen"
date: "2024-03-24"
output: html_document
---

# Set Up

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load libraries}
library(tidyverse)
library(devtools)
library(edgeR)
```

```{r read files}
sdiv <- read_tsv("../input/Christina_HiFi_Sdiv_Gene_Counts.tsv")
sample <- read_csv("../input/Hydrothermal_Round_1_Sample_Descriptions.csv")
```

## Environment Information

- sdiv = counts tsv file that was given, and later with non-normalized removed
- sample = sample cvs file that was given
- wet_sample = sample csv with only samples that are the condition we want
- wet_sdiv = counts for the condition we want
- wet_sdiv_matrix = matrix of counts
- dge_data = DGEList()
- wet_sdiv_normal = extracted normalized data
- wet_sdiv_normal_log = extracted normalized data (log)
- design = matrix of design conditions
- fit = glmFit of dge_data and design
- wet_lrt = glmLRT of fit and coefficient of all populations
- DEgene_wet = genes with a FDR < 0.01
- DEgene_wet_all = all genes

## Data

```{r retaining specific wet samples}
wet_sample <- sample %>%
  filter(condition == "24 hr 20˚ L wp0") %>%
  filter(sample != "i9x") %>%
  filter(sample != "i9b")
wet_sample
```

```{r retaining wet sdiv ones that have been normalized}
wet_sdiv <- sdiv %>%
  select(1, wet_sample$sample) %>%
  rename("gene_id" = 1) %>%
  select(sort(colnames(.))) %>%
  select(gene_id, everything())
wet_sdiv
```

```{r checking names match}
names(wet_sdiv[,-1]) == wet_sample$sample
```

## Cor Table

```{r doing cor()}
wet_sdiv %>%
  select(-gene_id) %>%
  cor() %>%
  gplots::heatmap.2(Rowv=FALSE, Colv=FALSE, dendrogram="none", trace = "none", col=viridis::viridis(25, begin=.25), margins=c(7,8))
# a and b are highly correlated
# i, j, and k are odd
```

## Calculate Normalization Factors

```{r making a matrix}
wet_sdiv_matrix <- wet_sdiv %>%
  select(-gene_id) %>%
  as.matrix()
rownames(wet_sdiv_matrix) <- wet_sdiv$gene_id
```

```{r calculating normalization factors}
dge_data <- DGEList(counts = wet_sdiv_matrix,
                    group = wet_sample$group,
                    samples = wet_sample)

dge_data <- calcNormFactors(dge_data, method = "TMM")
dge_data$samples
```

## Plotting BCV

```{r plotting BCV}
plotMDS(dge_data)
# a and b overlap again
```

## Extract Normalized Data

```{r extract normalized data}
wet_sdiv_normal <- cpm(dge_data)
wet_sdiv_normal_log <- cpm(dge_data, log = T)
```

```{r boxplots of normalized data}
boxplot(wet_sdiv_normal)
boxplot(wet_sdiv_normal_log)
# odd ones: i07, j06, j09, j10, k07, l06
```

## Removal of Non-normalized

```{r removing the non-normalized}
to_remove <- c("i07", "j06", "j09", "j10", "k07", "l06", "d07", "g09", "i08", "i10", "k09", "l09", "m07")
# looks like john retained l06
# added some more to be removed: "d07", "g09", "i08", "i10", "k09", "l09", "m07"
# john retained all of these
wet_sample <- wet_sample %>%
  filter(!(sample %in% to_remove))
wet_sample
```

```{r removing the not-normalized too}
wet_sdiv <- wet_sdiv %>%
  select(-to_remove)
wet_sdiv
```

```{r checking names match again}
names(wet_sdiv[,-1]) == wet_sample$sample
```

## Cor Table Again

```{r doing cor() again}
wet_sdiv %>%
  select(-gene_id) %>%
  cor() %>%
  gplots::heatmap.2(Rowv=FALSE, Colv=FALSE, dendrogram="none", trace = "none", col=viridis::viridis(25, begin=.25), margins=c(7,8))
# a and b are highly correlated
# i, j, and k area is still odd, but some of them were removed
```

## Calculate Normalization Factors Again

```{r making a matrix again}
wet_sdiv_matrix <- wet_sdiv %>%
  select(-gene_id) %>%
  as.matrix()
rownames(wet_sdiv_matrix) <- wet_sdiv$gene_id
```

```{r calculating normalization factors again}
dge_data <- DGEList(counts = wet_sdiv_matrix,
                    group = wet_sample$group,
                    samples = wet_sample)

dge_data <- calcNormFactors(dge_data, method = "TMM")
dge_data$samples
```

## Plotting BCV again

```{r plotting BCV again}
plotMDS(dge_data)
```

## Extract Normalized Data Again

```{r extract normalized data again}
wet_sdiv_normal <- cpm(dge_data)
wet_sdiv_normal_log <- cpm(dge_data, log = T)
```

```{r boxplots of normalized data again}
boxplot(wet_sdiv_normal)
boxplot(wet_sdiv_normal_log)
# more variation
# pulled more since we have so many replicates, may have more confusion/errors by keeping them in. results due to poor rna extraction
```

## Calculate Dispersion Factors & Estimate Dispersions

```{r make a model matrix}
design <- model.matrix(~ population, data = wet_sample)
rownames(design) <- wet_sample$sample.description
design
```

```{r estimate dispersion: common}
dge_data <- estimateGLMCommonDisp(dge_data, design, verbose = TRUE)
# Disp = 0.18392 , BCV = 0.4289
# Disp = 0.1024 , BCV = 0.32 
```

```{r estimate dispersion: trended}
dge_data <- estimateGLMTrendedDisp(dge_data, design)
```

```{r estimate dispersion: tagwise}
dge_data <- estimateGLMTagwiseDisp(dge_data, design)
```

```{r plotting biological coefficent of variation and average log CPM}
plotBCV(dge_data)
```

## Find Differentially Expressed Genes

```{r full model}
fit <- glmFit(dge_data, design)
```

```{r finding differentially expressed genes again}
wet_lrt <- glmLRT(fit, coef = colnames(design[,-1]))
```

```{r view the differentially expressed genes again}
topTags(wet_lrt)
```

```{r summarise number of differentially expressed genes again}
summary(decideTestsDGE(wet_lrt, p.value=0.01))
```

```{r extract genes with a FDR < 0.01 (could also use 0.05) and save to a file}
DEgene_wet <- topTags(wet_lrt, n = Inf, p.value = 0.01)$table # what is n = Inf for
write.csv(DEgene_wet,"../output/DEgenes_24hr20L0.csv")
```

```{r or if you want to keep all of them}
DEgene_wet_all <- topTags(wet_lrt, n = Inf, p.value = 1)$table
write.csv(DEgene_wet_all,"../output/DEgenes_24hr20L0_all.csv")
```

## Box Plot Function

```{r box plot function}
plotDE <- function(genes, dge, sample.description) {
  require(ggplot2)
  tmp.data <- t(log2(cpm(dge[genes,])+1))
  tmp.data <- tmp.data %>%
    as.data.frame() %>%
    rownames_to_column("sample") %>%
    left_join(sample.description,by="sample")
  tmp.data <- tmp.data %>%
    pivot_longer(cols=starts_with("Sdiv"), values_to = "log2_cpm", names_to = "gene")
  pl <- ggplot(tmp.data,aes(x=population,y=log2_cpm))
  pl <- pl + facet_wrap( ~ gene)
  pl <- pl + ylab("log2(cpm)") + xlab("genotype")
  pl <- pl + geom_boxplot()
  pl + theme(axis.text.x  = element_text(angle=45, vjust=1,hjust=1))
}
```

## Looking at FLOEs

```{r are they significant}
DEgene_wet_all %>%
  rownames_to_column("rowname") %>%
  filter(rowname == "Sdiv_ptg000004l_1566-R" | 
         rowname == "Sdiv_ptg000005l_1254-R" | 
         rowname == "Sdiv_ptg000009l_0928-R" | 
         rowname == "Sdiv_ptg000010l_1994-R" | 
         rowname == "Sdiv_ptg000001l_2274-R" | 
         rowname == "Sdiv_ptg000013l_FLOE3-R")
```


```{r using box plot function}
# FLOE1
plotDE(c("Sdiv_ptg000004l_1566-R", "Sdiv_ptg000005l_1254-R"), dge_data, wet_sample)

# FLOE2
plotDE(c("Sdiv_ptg000009l_0928-R", "Sdiv_ptg000010l_1994-R"), dge_data, wet_sample)
# also looks very similar to dry samples

# FLOE3
plotDE(c("Sdiv_ptg000001l_2274-R", "Sdiv_ptg000013l_FLOE3-R"), dge_data, wet_sample)
# same case as dry for the CACO1 and CAIN2 populations
```

## Correlation

```{r reading germination data}
# this is the code given to me
germData  <- read_csv("../input/Dry_and_Wp0_Germination_Data.csv") %>%
  group_by(population) %>%
  mutate(cumulative_prop_germ = max(cumulative_prop_germ)) %>%
  ungroup() %>%
  filter(condition == "24 hr 20˚ L wp0") %>%
  select(sample, population, cumulative_prop_germ)
germData
```

```{r correlation between floe expression and germination}
wet_sdiv_normal_log2 <- log2(cpm(dge_data) + 1)

cor_floe2 <- wet_sdiv_normal_log2 %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  filter(rowname == "Sdiv_ptg000004l_1566-R" | 
         rowname == "Sdiv_ptg000005l_1254-R" | 
         rowname == "Sdiv_ptg000009l_0928-R" | 
         rowname == "Sdiv_ptg000010l_1994-R" | 
         rowname == "Sdiv_ptg000001l_2274-R" | 
         rowname == "Sdiv_ptg000013l_FLOE3-R") %>%
  group_by(rowname) %>%
  mutate(CAAN1 = mean(c(a06, a07, a08, a09, a10)),
         CAAN2 = mean(c(b06, b07, b08, b09, b10)),
         CACO1 = mean(c(c06, c07, c08, c09, c10)),
         CAIN2 = mean(c(d06, d08, d09, d10)),
         STBR3 = mean(c(e06, e07, e08, e09, e10)),
         STDI = mean(c(f06, f07, f08, f09, f10)),
         STDR2 = mean(c(g06, g07, g08, g10)),
         STGL1 = mean(c(h06, h07, h09, h10)),
         STGL2 = mean(c(i06)),
         STGL3 = mean(c(j07, j08)),
         STIN = mean(c(k08, k10)),
         STPO1 = mean(c(l07, l08, l10)),
         STTO_BH = mean(c(m06 ,m08, m09, m10))) %>%
  select(rowname, CAAN1, CAAN2, CACO1, CAIN2, STBR3, STDI, STDR2, STGL1, STGL2, STGL3, STIN, STPO1, STTO_BH) %>%
  pivot_longer(cols = CAAN1:STTO_BH,
               names_to = "population",
               values_to = "log2cpm") %>%
  rename(gene_id = rowname)

germData2 <- germData %>%
  group_by(population) %>%
  summarise(cumulative_prop_germ = mean(cumulative_prop_germ)) %>%
  mutate(population = str_replace(population, "-", "_"))

cor_floe2 <- cor_floe2 %>%
  full_join(germData2)
  
cor_floe2
```

```{r correlation between floe expression and germination}
cor_floe2 %>%
  ggplot(aes(x = cumulative_prop_germ, y = log2cpm, color = gene_id)) +
  geom_point() + 
  facet_wrap(~ gene_id, scales = "free_y") +
    geom_smooth(method = lm, alpha = 0)
```

```{r correlation with nested values}
cor_floe2 %>%
  ungroup() %>%
  nest(.by=gene_id) %>% # separate observations by rowname and put them into separate tibbles
  mutate(cor_test = map(data, ~ cor.test(.$log2cpm, .$cumulative_prop_germ) )) %>% # make a new column and use the command cor.test on two columns in the tibbles
  mutate(glance = map(cor_test,  broom::glance)) %>% # take the cor.test results and make them readable and put them in a tibble in a new column called "glance"
  unnest(glance) %>% # add the tibble from the column "glance" to the end of the tibble
  arrange(p.value)
```