---
title: "FLOE notes v3"
author: "Christina Chen"
date: "2024-03-24"
output: html_document
---

# Set Up

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load libraries}
library(tidyverse)
library(devtools)
library(edgeR)
library(shiny)
```

```{r read files}
sdiv <- read_tsv("../input/Christina_HiFi_Sdiv_Gene_Counts.tsv")
sample <- read_csv("../input/Hydrothermal_Round_1_Sample_Descriptions.csv")
```

## Environment Information

- sdiv = counts tsv file that was given, and later with non-normalized removed
- sample = sample cvs file that was given
- dry_sample = sample csv with only dry condition, and without i02 and i03 and non-normalized
- dry_sdiv = counts tsv with only dry condition
- dry_sdiv_matrix = made into a matrix for rownames
- dge_data = DGEList and calcNormFactors
- normal_dry_sdiv = cpm of dge_data
- normal_dry_sdiv_log = cpm of dge_data
- sample_names = names of samples with names(dry_sdiv)
- design = matrix of design conditions
- fit = glmFit of dge_data and design
- dry_lrt = glmLRT of fit and coefficient of all populations

# Week 4

```{r retaining dry sdiv ones that have been normalized}
dry_sdiv <- sdiv %>%
  select(1, contains( "01") | contains("02") | contains("03") | contains("04") | contains("05")) %>%
  rename("gene_id" = 1) %>%
  select(-a01, -b02, -b04, -c01, -c02, -c03, -c05, -d01, -d02, -d03, -d04, -e03, -g01, -g02, -g03, -g04, -g05, -j03, -k01, -k03, -k04, -k05, -l01, -l04) %>%
  select(sort(colnames(.))) %>%
  select(gene_id, everything())
dry_sdiv
# ended up figuring out that i could do column_to_rownames()
```

```{r retaining sample descriptions to match dry_sdiv}
dry_sample <- sample %>%
  filter(condition == "Dry") %>%
  filter(sample != "i02",
         sample != "i03",
         sample != "a01",
         sample != "b02",
         sample != "b04",
         sample != "c01",
         sample != "c02",
         sample != "c03",
         sample != "c05",
         sample != "d01",
         sample != "d02",
         sample != "d03",
         sample != "d04",
         sample != "e03",
         sample != "g01",
         sample != "g02",
         sample != "g03",
         sample != "g04",
         sample != "g05",
         sample != "j03",
         sample != "k01",
         sample != "k03",
         sample != "k04",
         sample != "k05",
         sample != "l01",
         sample != "l04")
dry_sample
# how to filter with a c()?
```

```{r checking names match}
names(dry_sdiv[,-1]) == dry_sample$sample
```

## Cor Table

```{r doing cor()}
dry_sdiv %>%
  select(-gene_id) %>%
  cor() %>%
  gplots::heatmap.2(Rowv=FALSE, Colv=FALSE, dendrogram="none", trace = "none", col=viridis::viridis(25, begin=.25), margins=c(7,8))
```

## Calculate Normalization Factrs

```{r making a matrix}
dry_sdiv_matrix <- dry_sdiv %>%
  select(-gene_id) %>%
  as.matrix()
rownames(dry_sdiv_matrix) <- dry_sdiv$gene_id
```

```{r calculating normalization factors}
dge_data <- DGEList(counts = dry_sdiv_matrix,
                    group = dry_sample$group,
                    samples = dry_sample)

dge_data <- calcNormFactors(dge_data, method = "TMM")
dge_data$samples
```

## Plotting BCV

```{r}
plotMDS(dge_data, method = "bcv")
# why does it look so different from v2?
```

## Extract Normalized Data

```{r extract normalized data}
dry_sdiv_normal <- cpm(dge_data)
dry_sdiv_normal_log <- cpm(dge_data, log = T)
```

```{r boxplots of normalized data}
boxplot(dry_sdiv_normal)
boxplot(dry_sdiv_normal_log)
```

## Calculate Dispersion Factors & Estimate Dispersions

```{r make a model matrix}
design <- model.matrix(~ population, data = dry_sample)
rownames(design) <- dry_sample$sample.description
design
```

```{r estimate dispersion: common}
dge_data <- estimateGLMCommonDisp(dge_data, design, verbose = TRUE)
# Disp = 0.24255 , BCV = 0.4925 
```

```{r estimate dispersion: trended}
dge_data <- estimateGLMTrendedDisp(dge_data, design)
```

```{r estimate dispersion: tagwise}
dge_data <- estimateGLMTagwiseDisp(dge_data, design)
```

```{r plotting biological coefficent of variation and average log CPM}
plotBCV(dge_data)
# trend looks better
```

## Find Differentially Expressed Genes

```{r full model}
fit <- glmFit(dge_data, design)
```

```{r finding differentially expressed genes again}
dry_lrt <- glmLRT(fit, coef = colnames(design[,-1]))
```

```{r view the differentially expressed genes again}
topTags(dry_lrt)
```

### What topTags Shows

- logFC = log2 fold-change in expression, so populationCAAN2 (gene 20946) has 1.31, which means that gene is expressed 2^1.31 (or 2.48) more. if it is  a negative, it is less.
      - between intercept and a model without each population?
- logCPM = average expression across all samples
- LR = likeihood ratio (L(full model)/L(small model))
- PValue = unadjusted p value
- FDR = false discovery rate (p value adjusted for multiple testing)

```{r summarise number of differentially expressed genes again}
summary(decideTestsDGE(dry_lrt, p.value=0.01))
```

```{r extract genes with a FDR < 0.01 (could also use 0.05) and save to a file}
DEgene_dry <- topTags(dry_lrt, n = Inf, p.value = 0.01)$table # what is n = Inf for
write.csv(DEgene_dry,"../output/DEgenes_dry.csv")
```

```{r or if you want to keep all of them}
DEgene_dry_all <- topTags(dry_lrt, n = Inf, p.value = 1)$table
write.csv(DEgene_dry_all,"../output/DEgenes_dry_all.csv")
```

## Box Plot Function

```{r box plot function}
plotDE <- function(genes, dge, sample.description) {
  require(ggplot2)
  tmp.data <- t(log2(cpm(dge[genes,])+1))
  tmp.data <- tmp.data %>%
    as.data.frame() %>%
    rownames_to_column("sample") %>%
    left_join(sample.description,by="sample")
  tmp.data <- tmp.data %>%
    pivot_longer(cols=starts_with("Sdiv"), values_to = "log2_cpm", names_to = "gene")
  pl <- ggplot(tmp.data,aes(x=population,y=log2_cpm))
  pl <- pl + facet_wrap( ~ gene)
  pl <- pl + ylab("log2(cpm)") + xlab("genotype")
  pl <- pl + geom_boxplot()
  pl + theme(axis.text.x  = element_text(angle=45, vjust=1,hjust=1))
}
```

## Looking at FLOEs

```{r are they significant}
DEgene_dry_all %>%
  rownames_to_column("rowname") %>%
  filter(rowname == "Sdiv_ptg000004l_1566-R" | rowname == "Sdiv_ptg000005l_1254-R" | rowname == "Sdiv_ptg000009l_0928-R" | rowname == "Sdiv_ptg000010l_1994-R" | rowname == "Sdiv_ptg000001l_2274-R" | rowname == "Sdiv_ptg000013l_FLOE3-R")
```


```{r using box plot function}
# FLOE1
plotDE(c("Sdiv_ptg000004l_1566-R", "Sdiv_ptg000005l_1254-R"), dge_data, dry_sample)
# FLOE2
plotDE(c("Sdiv_ptg000009l_0928-R", "Sdiv_ptg000010l_1994-R"), dge_data, dry_sample)
# FLOE3
plotDE(c("Sdiv_ptg000001l_2274-R", "Sdiv_ptg000013l_FLOE3-R"), dge_data, dry_sample)
```