---
title: "FLOE notes v3"
author: "Christina Chen"
date: "2024-03-24"
output: html_document
---

# Set Up

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load libraries}
library(tidyverse)
library(devtools)
library(edgeR)
```

```{r read files}
sdiv <- read_tsv("../input/Christina_HiFi_Sdiv_Gene_Counts.tsv")
sample <- read_csv("../input/Hydrothermal_Round_1_Sample_Descriptions.csv")
```

## Environment Information

- sdiv = counts tsv file that was given, and later with non-normalized removed
- sample = sample cvs file that was given
- wet_sample = sample csv with only samples that are the condition we want
- wet_sdiv = counts for the condition we want
- wet_sdiv_matrix = matrix of counts
- dge_data = DGEList()
- wet_sdiv_normal = extracted normalized data
- wet_sdiv_normal_log = extracted normalized data (log)
- design = matrix of design conditions
- fit = glmFit of dge_data and design
- wet_lrt = glmLRT of fit and coefficient of all populations
- DEgene_wet = genes with a FDR < 0.01
- DEgene_wet_all = all genes

## Data

```{r retaining specific wet samples}
wet_sample <- sample %>%
  filter(condition == "24 hr 20Ëš L wp-0.3") %>%
  filter(sample != "a12") %>%
  filter(sample != "k15")
wet_sample
```

```{r retaining wet sdiv ones that have been normalized}
wet_sdiv <- sdiv %>%
  select(1, wet_sample$sample) %>%
  rename("gene_id" = 1) %>%
  select(sort(colnames(.))) %>%
  select(gene_id, everything())
wet_sdiv
```

```{r checking names match}
names(wet_sdiv[,-1]) == wet_sample$sample
```

## Cor Table

```{r doing cor()}
wet_sdiv %>%
  select(-gene_id) %>%
  cor() %>%
  gplots::heatmap.2(Rowv=FALSE, Colv=FALSE, dendrogram="none", trace = "none", col=viridis::viridis(25, begin=.25), margins=c(7,8))
# i, j, and k are odd (again)
```

## Calculate Normalization Factors

```{r making a matrix}
wet_sdiv_matrix <- wet_sdiv %>%
  select(-gene_id) %>%
  as.matrix()
rownames(wet_sdiv_matrix) <- wet_sdiv$gene_id
```

```{r calculating normalization factors}
dge_data <- DGEList(counts = wet_sdiv_matrix,
                    group = wet_sample$group,
                    samples = wet_sample)

dge_data <- calcNormFactors(dge_data, method = "TMM")
dge_data$samples
```

## Plotting BCV

```{r plotting BCV}
plotMDS(dge_data)
# some oddities, and m is spread out
```

## Extract Normalized Data

```{r extract normalized data}
wet_sdiv_normal <- cpm(dge_data)
wet_sdiv_normal_log <- cpm(dge_data, log = T)
```

```{r boxplots of normalized data}
boxplot(wet_sdiv_normal)
boxplot(wet_sdiv_normal_log)
# odd ones: i11, k12, l13, m12, m13, m14, m15
```

## Removal of Non-normalized

```{r removing the non-normalized}
to_remove <- c("i11", "k12", "l13", "m12", "m13", "m14", "m15")
# nothing to compare to
wet_sample <- wet_sample %>%
  filter(!(sample %in% to_remove))
wet_sample
```

```{r removing the not-normalized too}
wet_sdiv <- wet_sdiv %>%
  select(-to_remove)
wet_sdiv
```

```{r checking names match again}
names(wet_sdiv[,-1]) == wet_sample$sample
```

## Cor Table Again

```{r doing cor() again}
wet_sdiv %>%
  select(-gene_id) %>%
  cor() %>%
  gplots::heatmap.2(Rowv=FALSE, Colv=FALSE, dendrogram="none", trace = "none", col=viridis::viridis(25, begin=.25), margins=c(7,8))
# i, j, and k area is still odd
```

## Calculate Normalization Factors Again

```{r making a matrix again}
wet_sdiv_matrix <- wet_sdiv %>%
  select(-gene_id) %>%
  as.matrix()
rownames(wet_sdiv_matrix) <- wet_sdiv$gene_id
```

```{r calculating normalization factors again}
dge_data <- DGEList(counts = wet_sdiv_matrix,
                    group = wet_sample$group,
                    samples = wet_sample)

dge_data <- calcNormFactors(dge_data, method = "TMM")
dge_data$samples
```

## Plotting BCV again

```{r plotting BCV again}
plotMDS(dge_data)
# most m were removed so it looks better
```

## Extract Normalized Data Again

```{r extract normalized data again}
wet_sdiv_normal <- cpm(dge_data)
wet_sdiv_normal_log <- cpm(dge_data, log = T)
```

```{r boxplots of normalized data again}
boxplot(wet_sdiv_normal)
boxplot(wet_sdiv_normal_log)
```

## Calculate Dispersion Factors & Estimate Dispersions

```{r make a model matrix}
design <- model.matrix(~ population, data = wet_sample)
rownames(design) <- wet_sample$sample.description
design
```

```{r estimate dispersion: common}
dge_data <- estimateGLMCommonDisp(dge_data, design, verbose = TRUE)
# Disp = 0.20702 , BCV = 0.455
```

```{r estimate dispersion: trended}
dge_data <- estimateGLMTrendedDisp(dge_data, design)
```

```{r estimate dispersion: tagwise}
dge_data <- estimateGLMTagwiseDisp(dge_data, design)
```

```{r plotting biological coefficent of variation and average log CPM}
plotBCV(dge_data)
```

## Find Differentially Expressed Genes

```{r full model}
fit <- glmFit(dge_data, design)
```

```{r finding differentially expressed genes again}
wet_lrt <- glmLRT(fit, coef = colnames(design[,-1]))
```

```{r view the differentially expressed genes again}
topTags(wet_lrt)
# FDR (false discovery rate) - corrects for pvalue by, 
# two species: measured expression of FLOE1, do a t-test (mean comparison), if there is nothing going on, the chance of getting a low pvalue is what we define the pvalue
# pvalue says there is 5% chance of error, FDR says 5% of those are really false.
```

```{r summarise number of differentially expressed genes again}
summary(decideTestsDGE(wet_lrt, p.value=0.01))
```

```{r extract genes with a FDR < 0.01 (could also use 0.05) and save to a file}
DEgene_wet <- topTags(wet_lrt, n = Inf, p.value = 0.01)$table # what is n = Inf for
write.csv(DEgene_wet,"../output/DEgenes_24hr20L03.csv")
# n for amount shown
# sorted by FDR or adjusted pvalue
```

```{r or if you want to keep all of them}
DEgene_wet_all <- topTags(wet_lrt, n = Inf, p.value = 1)$table
write.csv(DEgene_wet_all,"../output/DEgenes_24hr20L03_all.csv")
```

## Box Plot Function

```{r box plot function}
plotDE <- function(genes, dge, sample.description) {
  require(ggplot2)
  tmp.data <- t(log2(cpm(dge[genes,])+1))
  tmp.data <- tmp.data %>%
    as.data.frame() %>%
    rownames_to_column("sample") %>%
    left_join(sample.description,by="sample")
  tmp.data <- tmp.data %>%
    pivot_longer(cols=starts_with("Sdiv"), values_to = "log2_cpm", names_to = "gene")
  pl <- ggplot(tmp.data,aes(x=population,y=log2_cpm))
  pl <- pl + facet_wrap( ~ gene)
  pl <- pl + ylab("log2(cpm)") + xlab("genotype")
  pl <- pl + geom_boxplot()
  pl + theme(axis.text.x  = element_text(angle=45, vjust=1,hjust=1))
}
```

## Looking at FLOEs

```{r are they significant}
DEgene_wet_all %>%
  rownames_to_column("rowname") %>%
  filter(rowname == "Sdiv_ptg000004l_1566-R" | 
         rowname == "Sdiv_ptg000005l_1254-R" | 
         rowname == "Sdiv_ptg000009l_0928-R" | 
         rowname == "Sdiv_ptg000010l_1994-R" | 
         rowname == "Sdiv_ptg000001l_2274-R" | 
         rowname == "Sdiv_ptg000013l_FLOE3-R")
```


```{r using box plot function}
# FLOE1
plotDE(c("Sdiv_ptg000004l_1566-R", "Sdiv_ptg000005l_1254-R"), dge_data, wet_sample)

# FLOE2
plotDE(c("Sdiv_ptg000009l_0928-R", "Sdiv_ptg000010l_1994-R"), dge_data, wet_sample)
# also looks very similar to dry samples

# FLOE3
plotDE(c("Sdiv_ptg000001l_2274-R", "Sdiv_ptg000013l_FLOE3-R"), dge_data, wet_sample)
# nothing for ptg13
```